packages/@react-aria/focus/src/FocusScope.tsx

```diff
@@ -119,30 +119,46 @@ export function useFocusManager(): FocusManager {
 function createFocusManager(scopeRef: React.RefObject<HTMLElement[]>): FocusManager {
   return {
     focusNext(opts: FocusManagerOptions = {}) {
-      let node = opts.from || document.activeElement;
-      let focusable = getFocusableElementsInScope(scopeRef.current, opts);
-      let nextNode = focusable.find(n =>
-        !!(node.compareDocumentPosition(n) & (Node.DOCUMENT_POSITION_FOLLOWING | Node.DOCUMENT_POSITION_CONTAINED_BY))
-      );
-      if (!nextNode && opts.wrap) {
-        nextNode = focusable[0];
+      let scope = scopeRef.current;
+      let {from, tabbable, wrap} = opts; 
+      let node = from || document.activeElement;
+      let sentinel = scope[0].previousElementSibling;
+      let walker = getFocusableTreeWalker(getScopeRoot(scope), {tabbable});
+      walker.currentNode = isElementInScope(node, scope) ? node : sentinel;
+      let nextNode = walker.nextNode() as HTMLElement;
+      if (!(nextNode && isElementInScope(nextNode, scope)) && wrap) {
+        walker.currentNode = sentinel;
+        nextNode = walker.nextNode() as HTMLElement;
       }
-      if (nextNode) {
-        nextNode.focus();
+      if (nextNode && isElementInScope(nextNode, scope)) {
+        focusElement(nextNode, true);
       }
       return nextNode;
     },
     focusPrevious(opts: FocusManagerOptions = {}) {
-      let node = opts.from || document.activeElement;
-      let focusable = getFocusableElementsInScope(scopeRef.current, opts).reverse();
-      let previousNode = focusable.find(n =>
-        !!(node.compareDocumentPosition(n) & (Node.DOCUMENT_POSITION_PRECEDING | Node.DOCUMENT_POSITION_CONTAINED_BY))
-      );
-      if (!previousNode && opts.wrap) {
-        previousNode = focusable[0];
+      let scope = scopeRef.current;
+      let {from, tabbable, wrap} = opts; 
+      let node = from || document.activeElement;
+      let sentinel = scope[scope.length - 1].nextElementSibling;
+      let walker = getFocusableTreeWalker(getScopeRoot(scope), {tabbable});
+      walker.currentNode = sentinel;
+      let prevNode = walker.previousNode() as HTMLElement;
+      let previousNode;
+      while (prevNode && isElementInScope(prevNode, scope))  {
```

Comment:
so I tried your branch and replaced this section with the mirror of the above focusNext code, it works in all cases, except for two tests, one in DatePicker and one in DateRangePicker, if I had to guess, those two components need to be fixed, rather than this code needing to compensate for them. It's either that, or FocusScope.test.tsx needs to have a test case that highlights what that different use case is. I'm looking into what's happening in DatePicker now, but if you already know, I figured I'd comment here.

Reply:
Datepicker has the event hander `onMouseDown` in `useDateField`, which calls `focusManager.focusPrevious` with the dateField element as the `from` prop. The Datepicker test expects the last segment element within the dateField that handles the mouse down event to receive focus.  However the the dateField container passed as the `from` element is not focusable and has no preceding focusable elements, so without calling  `compareDocumentPosition` to locate a segment before or contained within the `from` element, the FocusScope will simply wrap focus to the last item within the scope. For a DateRangePicker, this could be the last segment in a dateField other than the one that triggered `onMouseDown`.

Reply:
Note that with `focusNext`, I don't think this is a problem, because focusable elements contained within a `from` element will always follow the `from` element when tree walking.

Reply:
should DatePicker actually call `focusNext` or maybe call focus on the last field directly instead of using the focusManager? or why must it call `focusPrevious`?

I'm actually playing with the storybook built from main, I'm unable to click into a DatePicker at all...

I guess what i'm getting at is, what does it mean to be calling focusPrevious... is this actually a real use case?

@devongovett ?

Reply:
I'm not sure that this is a real use case. My PR to address Datepicker issues, https://github.com/adobe/react-spectrum/pull/246, corrects the behavior of onPressStart to set focus to date segments, and in my testing within Storybook, I haven't been able to get the onMouseDown event within useDateField to ever actually get called.

Reply:
Then my opinion is that we delete the test as it isn't valid, I haven't been able to fire that event either, and we simplify our logic so forward and backward match. Especially since your other PR addresses it. I'll ping devon directly too to make sure he's ok with it.